xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -134,11 +142,318 @@ func (s) TestRingHash_ReconnectToMoveOutOfTransientFailure(t *testing.T) {
 		}
 	}
 	if err := ctx.Err(); err != nil {
-		t.Fatalf("Timeout waiting for channel to reach READT after server restart: %v", err)
+		t.Fatalf("Timeout waiting for channel to reach READY after server restart: %v", err)
 	}
 
 	// An RPC at this point is expected to fail.
 	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
 		t.Fatalf("rpc EmptyCall() failed: %v", err)
 	}
 }
+
+// XXX this is copy pasted from
+// https://github.com/grpc/grpc-go/blob/9199290ff8fee7c37283586929af39ba23dbab4b/xds/internal/balancer/clusterresolver/e2e_test/eds_impl_test.go#L93
+// Perhaps we should move it to a common package?
```

Comment:
We *probably could* move it to a common package. I would assume then that the StubServer would need to implement all methods on the `TestService` to some acceptable degree.

But given that this is already going to a big change, I'm totally OK with having some repetitive test helper code spread across our code base.